sound/soc/sof/pm.c

```diff
@@ -425,3 +425,46 @@ int snd_sof_suspend(struct device *dev)
 	return sof_suspend(dev, false);
 }
 EXPORT_SYMBOL(snd_sof_suspend);
+
+/*
+ * DSP D0 sub-states(d0i0/d0i3) may transform as below:-
+ *
+ *       -------------------
+ *       |                 |<------------------------+
+ *  +--->|       D0I0      |                         |
+ *  |    |                 |<---------------+        |
+ *  |    -------------------                |        |
+ *  |         ||d0i3 exit |                 |        |
+ *  | A normal||forcing   | A low power     |        |
+ *  |substream||task done,| substream   Yes |        |
+ *  | closed  ||return    | opened          |        |A normal
```

Comment:
ok, so my suggestion is to simplify the state machine by labeling the arcs, e.g. (1) (2) etc, and have test that describe what causes each transition. It's too hard to read otherwise and it's hard to extend.